pydicom/pixel_data_handlers/pillow_handler.py

```diff
@@ -89,49 +90,130 @@ def should_change_PhotometricInterpretation_to_RGB(ds: "Dataset") -> bool:
+        return im
+
+    cs = None
+
+    # Parse the JPEG codestream for the APP and SOF markers
+    param = parse_jpeg(src)
+
+    # APP0 JFIF implies YCbCr
+    # https://www.w3.org/Graphics/JPEG/jfif3.pdf
+    if "APPn" in param:
+        if param["APPn"].get(b"\xFF\xE0", b"").startswith(b"JFIF"):
+            cs = "YCbCr"
+
+    # ISO/IEC 10918-1 (ITU T.81)
+    # https://www.w3.org/Graphics/JPEG/itu-t81.pdf
+    if "SOF" in param:
+        # If any components are subsampled then it's very likely not RGB
+        c_ss = [x[1] != 1 or x[2] != 1 for x in param["SOF"]["Components"]]
+        if any(c_ss):
+            cs = "YCbCr"
+
+        # Some applications use the SOF marker's component IDs to flag
+        #   the colour components
+        c_ids = [x[0] for x in param["SOF"]["Components"]]
+        if c_ids in ([b"R", b"G", b"B"], [b"r", b"g", b"b"]):
+            cs = "RGB"
+
+    # APP14 (Adobe)
+    # ISO/IEC 10918-6 (ITU T.872): JPEG Printing Extensions
+    # http://www.itu.int/rec/T-REC-T.872-201206-I/en
+    # Section 6.5.3:
+    #   AP12 is assumed to contain a single-byte transform flag as:
+    #     0: CMYK (if 4 components) or RGB (if 3 components)
+    #     1: YCbCr (3 components)
+    #     2: YCCK (4 components)
+    # In Image.info, "adobe_transform" is the APP14 AP12 value as int
+    if "adobe_transform" in im.info:
+        cs = "RGB" if im.info["adobe_transform"] == 0 else "YCbCr"
+
+    # We want to return the pixel data in the color space specified by
+    #   *Photometric Interpretation*, with a warning if that color space
+    #   doesn't actually match that of the encoded source data
+
+    # The possibilities are:
+    #   cs    | PI  -> transform    | return | warn
+    #   ------+---------------------+--------+-----
+    #   None  | RGB -> YCbCr to RGB | RGB    |
+    #   YCbCr | RGB -> YCbCr to RGB | RGB    | yes
+    #   RGB   | RGB -> (none)       | RGB    |
+    #   None  | YBR -> (none)       | YBR    |
+    #   YCbCr | YBR -> (none)       | YBR    |
+    #   RGB   | YBR -> RGB to YCbCr | YBR    | yes
+
+    if photometric_interpretation == "RGB":
+        if cs == "RGB":
+            # Source data is RGB - allow libjpeg to skip transform
+            # If "adobe_transform" is present, or if the component IDs are
+            #   RGB then libjpeg will use that info to skip the YCbCr -> RGB
+            #   transform
+            # https://github.com/libjpeg-turbo/ijg/blob/main/jdapimin.c
+            pass
+
+        if cs is None:
+            # Source data may be YCbCr (most likely) or RGB (less likely)
+            # libjpeg assumes source is YCbCr - force no transform
+            # Assume *Photometric Interpretation* is correct:
+            # * If the decoded pixel data is correct then source is RGB
+            # * If the decoded pixel data is incorrect then source is YCbCr
+            #   but this can be fixed by user applying YCbCr -> RGB transform
```

Comment:
Assuming the *Photometric Interpretation* is incorrect then we'd let libjpeg apply the YCbCr -> RGB transform, which gives:
* If the decoded pixel data is correct, source is YCbCr
* If the decoded pixel data is incorrect, source is RGB and the user would then need to apply RGB -> YCbCr transform to return to RGB (which is non-intuitive)

Reply:
Should we add a warning here? 

> If the returned pixel data is not in the RGB colour space then you will need to convert from YCbCr -> RGB using 'pydicom.pixel_data_handlers.convert_color_space()'

Reply:
Yes, I agree. Something like this was the case with the recent question, which confused me, because I assumed it would be the other way (as you are doing it now, as I understand). 
I'm not really able to review this, would have to do some readup on JPEG first, but I won't have much time for a week due to family stuff. The concept looks fine to me, though, from what I can understand.

Reply:
I would not make assumptions about the value of Photometric Interpretation being incorrect. If the value is `"RGB"`, it means that the data has been compressed in RGB color space without transformation into YCbCr color space and the decompressed data shall therefore not be color transformed.

Reply:
If there is a mismatch between the information in the JPEG codestream and the DICOM metadata, I tend argue that an error should be raised instead of a warning being logged. If necessary, the caller can work around such issues by correcting the DICOM metadata before decompression.

Reply:
> I would not make assumptions about the value of Photometric Interpretation being incorrect. 

I haven't though, I've assumed it is correct (which is also the logic currently used).

I was going to raise on mismatches, but I think that YCbCr is much more likely then unflagged RGB, so I was worried about how frequently we would get "I can't decode my pixel data" and similar issues.

Reply:
My comment is related to the following lines:

```python
    if photometric_interpretation == "RGB":
        ...
        if cs is None:
            # Source data may be YCbCr (most likely) or RGB (less likely)
```

If the value of Photometric Interpretation is `"RGB"`, than no color transformation should be applied during decoding. If for whatever reason the value of the data element is wrong, than that's not something the library should try to be smart and work around in my opinion, since it could lead to a wrong result. We have to be able to rely on the correctness of the metadata in the DICOM image.

Reply:
I agree